File: dspark/curator.py
"""
DeepSeek Curator & Arbitrator Engine for DSpark.
"""

from dataclasses import dataclass, field
from enum import Enum
import json
import re
from typing import Any, Dict, List, Optional, Union

from .client import DeepSeekClient
from .prompts import CURATOR_SYSTEM_PROMPT, ARBITRATOR_SYSTEM_PROMPT, REFINER_SYSTEM_PROMPT


class CurationVerdict(str, Enum):
    APPROVED = "APPROVED"
    NEEDS_REVISION = "NEEDS_REVISION"
    REJECTED = "REJECTED"


@dataclass
class EdgeCase:
    case: str
    risk_level: str
    handled_properly: bool
    remedy: str = ""


@dataclass
class CounterExample:
    failing_input: str
    expected_behavior: str
    actual_behavior: str
    severity: str = "HIGH"


@dataclass
class AuditResult:
    verdict: CurationVerdict
    score: int
    summary: str
    criteria_scores: Dict[str, int] = field(default_factory=dict)
    counter_examples: List[CounterExample] = field(default_factory=list)
    io_contract_analysis: Dict[str, Any] = field(default_factory=dict)
    edge_cases: List[EdgeCase] = field(default_factory=list)
    complexity: Dict[str, Any] = field(default_factory=dict)
    critical_issues: List[str] = field(default_factory=list)
    suggested_improvements: List[str] = field(default_factory=list)
    refined_code: Optional[str] = None
    raw_response: str = ""

    @property
    def is_approved(self) -> bool:
        return self.verdict == CurationVerdict.APPROVED


@dataclass
class RefineResult:
    refined_code: str
    summary_of_changes: List[str] = field(default_factory=list)
    raw_response: str = ""


@dataclass
class ArbitrationResult:
    winner_index: int
    rationale: str
    comparison_matrix: Dict[str, Any] = field(default_factory=dict)
    synthesized_code: str = ""
    raw_response: str = ""


def _extract_json(text: str) -> Dict[str, Any]:
    """Helper to extract JSON object from markdown blocks or free text."""
    text = text.strip()
    
    # Strip <think>...</think> tags if present
    text = re.sub(r"<think>.*?</think>", "", text, flags=re.DOTALL).strip()

    # 1. Try direct parse with strict=False
    try:
        return json.loads(text, strict=False)
    except Exception:
        pass

    # 2. Try extracting inside ```json ... ```
    match = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", text, re.DOTALL)
    if match:
        try:
            return json.loads(match.group(1), strict=False)
        except Exception:
            pass

    # 3. Try finding outermost { and }
    first_brace = text.find("{")
    last_brace = text.rfind("}")
    if first_brace != -1 and last_brace != -1 and last_brace > first_brace:
        candidate = text[first_brace:last_brace + 1]
        try:
            return json.loads(candidate, strict=False)
        except Exception:
            pass

    # 4. Fallback: repair unescaped newlines inside strings
    def _clean_newlines(m):
        return m.group(0).replace("\n", "\\n").replace("\r", "\\r")
    
    try:
        cleaned = re.sub(r'"([^"\\]*(\\.[^"\\]*)*)"', _clean_newlines, text)
        first_b = cleaned.find("{")
        last_b = cleaned.rfind("}")
        if first_b != -1 and last_b != -1:
            return json.loads(cleaned[first_b:last_b + 1], strict=False)
    except Exception:
        pass

    raise ValueError(f"Failed to parse valid JSON from model response:\n{text[:400]}...")


def _extract_code_blocks(text: str) -> str:
    """Extract code from triple backtick blocks if present."""
    match = re.search(r"```(?:\w+)?\n(.*?)```", text, re.DOTALL)
    if match:
        return match.group(1).strip()
    return text.strip()


class DeepSeekCurator:
    """
    Curator and Arbitrator engine powered by DeepSeek Reasoning, OpenAI, or Local LLMs.
    """

    def __init__(self, client: Optional[Any] = None, model: Optional[str] = None):
        if client:
            self.client = client
            self.model = model or getattr(client, "default_model", None)
        elif model:
            from .client import create_model_client
            self.client = create_model_client(model)
            self.model = getattr(self.client, "default_model", model)
        else:
            self.client = DeepSeekClient()
            self.model = self.client.default_model

    def audit(
        self,
        code: str,
        specification: str,
        language: Optional[str] = None,
        temperature: float = 0.1,
    ) -> AuditResult:
        """
        Audit draft code against a specification, validating I/O contracts and edge cases.
        """
        lang_str = f"Language: {language}\n" if language else ""
        user_prompt = (
            f"### SPECIFICATION / REQUIREMENTS:\n{specification}\n\n"
            f"### CANDIDATE IMPLEMENTATION TO AUDIT:\n{lang_str}```{language or ''}\n{code}\n```\n\n"
            f"Perform strict reasoning audit, I/O verification, edge case simulation, and return the required JSON."
        )

        raw_resp = self.client.complete(
            prompt=user_prompt,
            system_prompt=CURATOR_SYSTEM_PROMPT,
            model=self.model,
            temperature=temperature,
            response_format={"type": "json_object"},
        )

        data = _extract_json(raw_resp)
        verdict_str = str(data.get("verdict", "NEEDS_REVISION")).upper()
        verdict = CurationVerdict(verdict_str) if verdict_str in CurationVerdict.__members__ else CurationVerdict.NEEDS_REVISION

        edge_cases = []
        for ec in data.get("edge_cases_identified", []):
            if isinstance(ec, dict):
                edge_cases.append(
                    EdgeCase(
                        case=ec.get("case", ""),
                        risk_level=ec.get("risk_level", "MEDIUM"),
                        handled_properly=bool(ec.get("handled_properly", False)),
                        remedy=ec.get("remedy", ""),
                    )
                )

        counter_examples = []
        for ce in data.get("counter_examples", []):
            if isinstance(ce, dict):
                counter_examples.append(
                    CounterExample(
                        failing_input=str(ce.get("failing_input", "")),
                        expected_behavior=str(ce.get("expected_behavior", "")),
                        actual_behavior=str(ce.get("actual_behavior", "")),
                        severity=str(ce.get("severity", "HIGH")),
                    )
                )

        refined = data.get("refined_code", "").strip() or None

        return AuditResult(
            verdict=verdict,
            score=int(data.get("score", 70)),
            summary=data.get("summary", ""),
            criteria_scores=data.get("criteria_scores", {}),
            counter_examples=counter_examples,
            io_contract_analysis=data.get("io_contract_analysis", {}),
            edge_cases=edge_cases,
            complexity=data.get("complexity", {}),
            critical_issues=data.get("critical_issues", []),
            suggested_improvements=data.get("suggested_improvements", []),
            refined_code=refined,
            raw_response=raw_resp,
        )

    def refine(
        self,
        code: str,
        specification: str,
        feedback: Optional[str] = None,
        language: Optional[str] = None,
        temperature: float = 0.2,
    ) -> RefineResult:
        """
        Produce a refined, production-grade version of the code.
        """
        feedback_section = f"### AUDIT FEEDBACK / ISSUES TO FIX:\n{feedback}\n\n" if feedback else ""
        lang_str = f"Language: {language}\n" if language else ""
        user_prompt = (
            f"### SPECIFICATION:\n{specification}\n\n"
            f"{feedback_section}"
            f"### DRAFT CODE:\n{lang_str}```{language or ''}\n{code}\n```\n\n"
            f"Refine the code to 100% production readiness. Fix all potential edge cases and enforce strict I/O typing."
        )

        raw_resp = self.client.complete(
            prompt=user_prompt,
            system_prompt=REFINER_SYSTEM_PROMPT,
            model=self.model,
            temperature=temperature,
        )

        refined_code = _extract_code_blocks(raw_resp)
        
        # Extract bullets if present
        changes = []
        for line in raw_resp.splitlines():
            line = line.strip()
            if line.startswith(("- ", "* ", "• ")) and not line.startswith("```"):
                changes.append(line[2:].strip())

        return RefineResult(
            refined_code=refined_code,
            summary_of_changes=changes,
            raw_response=raw_resp,
        )

    def arbitrate(
        self,
        candidates: List[str],
        specification: str,
        language: Optional[str] = None,
        temperature: float = 0.1,
    ) -> ArbitrationResult:
        """
        Arbitrate between two or more candidate code implementations.
        """
        if len(candidates) < 2:
            raise ValueError("Arbitration requires at least 2 candidate implementations.")

        candidates_formatted = []
        for idx, cand in enumerate(candidates):
            candidates_formatted.append(
                f"### CANDIDATE #{idx}:\n```{language or ''}\n{cand}\n```"
            )

        user_prompt = (
            f"### SPECIFICATION:\n{specification}\n\n"
            + "\n\n".join(candidates_formatted)
            + "\n\nCompare the candidates thoroughly, choose the winner and synthesize the optimal code in the specified JSON format."
        )

        raw_resp = self.client.complete(
            prompt=user_prompt,
            system_prompt=ARBITRATOR_SYSTEM_PROMPT,
            model=self.model,
            temperature=temperature,
            response_format={"type": "json_object"},
        )

        data = _extract_json(raw_resp)
        return ArbitrationResult(
            winner_index=int(data.get("winner_index", 0)),
            rationale=data.get("rationale", ""),
            comparison_matrix=data.get("comparison_matrix", {}),
            synthesized_code=data.get("synthesized_code", ""),
            raw_response=raw_resp,
        )

File: src/curator.rs
//! DeepSeek Curator: audit, refine, and arbitrate implementations.

use crate::client::{ClientError, ModelClient};
use crate::prompts::{ARBITRATOR_SYSTEM_PROMPT, REFINER_SYSTEM_PROMPT};
use crate::verifier::VERIFIER_SYSTEM_PROMPT;
use crate::util::{extract_code_blocks, extract_json, json_string, json_string_vec, json_u32};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CuratorError {
    #[error("Client error: {0}")]
    Client(#[from] ClientError),
    #[error("{0}")]
    Parse(String),
    #[error("{0}")]
    Invalid(String),
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum CurationVerdict {
    Approved,
    NeedsRevision,
    Rejected,
}

impl CurationVerdict {
    pub fn from_str_loose(raw: &str) -> Self {
        match raw.to_uppercase().as_str() {
            "APPROVED" => Self::Approved,
            "REJECTED" => Self::Rejected,
            _ => Self::NeedsRevision,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Approved => "APPROVED",
            Self::NeedsRevision => "NEEDS_REVISION",
            Self::Rejected => "REJECTED",
        }
    }
}

impl std::fmt::Display for CurationVerdict {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct EdgeCase {
    pub case: String,
    pub risk_level: String,
    pub handled_properly: bool,
    #[serde(default)]
    pub remedy: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct CounterExample {
    pub failing_input: String,
    pub expected_behavior: String,
    pub actual_behavior: String,
    #[serde(default = "default_severity")]
    pub severity: String,
}

fn default_severity() -> String {
    "HIGH".to_string()
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct ComplexityAnalysis {
    #[serde(default)]
    pub time: String,
    #[serde(default)]
    pub space: String,
    #[serde(default)]
    pub optimal: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AuditResult {
    pub verdict: CurationVerdict,
    pub score: u32,
    pub summary: String,
    #[serde(default)]
    pub criteria_scores: Value,
    #[serde(default)]
    pub counter_examples: Vec<CounterExample>,
    #[serde(default)]
    pub io_contract_analysis: Value,
    #[serde(default)]
    pub edge_cases: Vec<EdgeCase>,
    #[serde(default)]
    pub complexity: Option<ComplexityAnalysis>,
    #[serde(default)]
    pub critical_issues: Vec<String>,
    #[serde(default)]
    pub suggested_improvements: Vec<String>,
    #[serde(default)]
    pub refined_code: Option<String>,
    #[serde(default)]
    pub raw_response: String,
}

impl AuditResult {
    pub fn is_approved(&self) -> bool {
        self.verdict == CurationVerdict::Approved
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RefineResult {
    pub refined_code: String,
    #[serde(default)]
    pub summary_of_changes: Vec<String>,
    #[serde(default)]
    pub raw_response: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ArbitrationResult {
    pub winner_index: i64,
    pub rationale: String,
    #[serde(default)]
    pub comparison_matrix: Value,
    #[serde(default)]
    pub synthesized_code: String,
    #[serde(default)]
    pub raw_response: String,
}

pub struct DeepSeekCurator {
    client: ModelClient,
}

impl DeepSeekCurator {
    pub fn new() -> Result<Self, CuratorError> {
        Ok(Self {
            client: ModelClient::from_spec(
                &std::env::var("DEEPSEEK_MODEL").unwrap_or_else(|_| "deepseek-v4-flash".into()),
            )?,
        })
    }

    pub fn with_model(spec: &str) -> Result<Self, CuratorError> {
        Ok(Self {
            client: ModelClient::from_spec(spec)?,
        })
    }

    pub fn with_client(client: ModelClient) -> Self {
        Self { client }
    }

    async fn complete_json(
        &self,
        prompt: &str,
        system: &str,
        temperature: f32,
    ) -> Result<String, CuratorError> {
        match self
            .client
            .complete(prompt, Some(system), temperature, true)
            .await
        {
            Ok(s) => Ok(s),
            Err(_) => Ok(self
                .client
                .complete(prompt, Some(system), temperature, false)
                .await?),
        }
    }

    pub async fn audit(
        &self,
        code: &str,
        specification: &str,
        language: Option<&str>,
    ) -> Result<AuditResult, CuratorError> {
        let lang_str = language
            .map(|l| format!("Language: {}\n", l))
            .unwrap_or_default();
        let lang_fence = language.unwrap_or("");
        let user_prompt = format!(
            "### SPECIFICATION / REQUIREMENTS:\n{specification}\n\n\
             ### CANDIDATE IMPLEMENTATION TO AUDIT:\n{lang_str}```{lang_fence}\n{code}\n```\n\n\
             Perform strict reasoning audit, I/O verification, edge case simulation, and return the required JSON."
        );

        let raw_resp = self
            .complete_json(&user_prompt, VERIFIER_SYSTEM_PROMPT, 0.1)
            .await?;
        let data = extract_json(&raw_resp).map_err(CuratorError::Parse)?;
        Ok(audit_from_value(&data, raw_resp))
    }

    pub async fn refine(
        &self,
        code: &str,
        specification: &str,
        feedback: Option<&str>,
        language: Option<&str>,
    ) -> Result<RefineResult, CuratorError> {
        let feedback_section = feedback
            .map(|f| format!("### AUDIT FEEDBACK / ISSUES TO FIX:\n{f}\n\n"))
            .unwrap_or_default();
        let lang_str = language
            .map(|l| format!("Language: {}\n", l))
            .unwrap_or_default();
        let lang_fence = language.unwrap_or("");
        let user_prompt = format!(
            "### SPECIFICATION:\n{specification}\n\n{feedback_section}\
             ### DRAFT CODE:\n{lang_str}```{lang_fence}\n{code}\n```\n\n\
             Refine the code to 100% production readiness. Fix all potential edge cases and enforce strict I/O typing."
        );

        let raw_resp = self
            .client
            .complete(&user_prompt, Some(REFINER_SYSTEM_PROMPT), 0.2, false)
            .await?;
        let refined_code = extract_code_blocks(&raw_resp);

        let changes: Vec<String> = raw_resp
            .lines()
            .map(str::trim)
            .filter(|line| {
                (line.starts_with("- ") || line.starts_with("* ") || line.starts_with("• "))
                    && !line.starts_with("```")
            })
            .map(|line| line.chars().skip(2).collect::<String>().trim().to_string())
            .collect();

        Ok(RefineResult {
            refined_code,
            summary_of_changes: changes,
            raw_response: raw_resp,
        })
    }

    pub async fn arbitrate(
        &self,
        candidates: &[String],
        specification: &str,
        language: Option<&str>,
    ) -> Result<ArbitrationResult, CuratorError> {
        if candidates.len() < 2 {
            return Err(CuratorError::Invalid(
                "Arbitration requires at least 2 candidate implementations.".into(),
            ));
        }

        let lang_fence = language.unwrap_or("");
        let mut body = format!("### SPECIFICATION:\n{specification}\n\n");
        for (idx, cand) in candidates.iter().enumerate() {
            body.push_str(&format!(
                "### CANDIDATE #{idx}:\n```{lang_fence}\n{cand}\n```\n\n"
            ));
        }
        body.push_str(
            "Compare the candidates thoroughly, choose the winner and synthesize the optimal code in the specified JSON format.",
        );

        let raw_resp = self
            .complete_json(&body, ARBITRATOR_SYSTEM_PROMPT, 0.1)
            .await?;
        let data = extract_json(&raw_resp).map_err(CuratorError::Parse)?;

        let winner_index = data
            .get("winner_index")
            .and_then(|v| v.as_i64().or_else(|| v.as_u64().map(|n| n as i64)))
            .unwrap_or(0);

        Ok(ArbitrationResult {
            winner_index,
            rationale: json_string(&data, "rationale"),
            comparison_matrix: data
                .get("comparison_matrix")
                .cloned()
                .unwrap_or(Value::Object(Default::default())),
            synthesized_code: json_string(&data, "synthesized_code"),
            raw_response: raw_resp,
        })
    }
}

fn audit_from_value(data: &Value, raw: String) -> AuditResult {
    let verdict = CurationVerdict::from_str_loose(&json_string(data, "verdict"));

    let mut edge_cases = Vec::new();
    let edge_src = data
        .get("edge_cases_identified")
        .or_else(|| data.get("edge_cases"))
        .and_then(|v| v.as_array());
    if let Some(arr) = edge_src {
        for ec in arr {
            edge_cases.push(EdgeCase {
                case: json_string(ec, "case"),
                risk_level: ec
                    .get("risk_level")
                    .and_then(|v| v.as_str())
                    .unwrap_or("MEDIUM")
                    .to_string(),
                handled_properly: ec
                    .get("handled_properly")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false),
                remedy: json_string(ec, "remedy"),
            });
        }
    }

    let mut counter_examples = Vec::new();
    if let Some(arr) = data.get("counter_examples").and_then(|v| v.as_array()) {
        for ce in arr {
            counter_examples.push(CounterExample {
                failing_input: json_string(ce, "failing_input"),
                expected_behavior: json_string(ce, "expected_behavior"),
                actual_behavior: json_string(ce, "actual_behavior"),
                severity: ce
                    .get("severity")
                    .and_then(|v| v.as_str())
                    .unwrap_or("HIGH")
                    .to_string(),
            });
        }
    }

    let complexity = data.get("complexity").map(|c| ComplexityAnalysis {
        time: json_string(c, "time"),
        space: json_string(c, "space"),
        optimal: c.get("optimal").and_then(|v| v.as_bool()).unwrap_or(false),
    });

    let refined = json_string(data, "refined_code");
    let refined_code = if refined.trim().is_empty() {
        None
    } else {
        Some(refined)
    };

    AuditResult {
        verdict,
        score: json_u32(data, "score", 70),
        summary: json_string(data, "summary"),
        criteria_scores: data
            .get("criteria_scores")
            .cloned()
            .unwrap_or(Value::Object(Default::default())),
        counter_examples,
        io_contract_analysis: data
            .get("io_contract_analysis")
            .cloned()
            .unwrap_or(Value::Object(Default::default())),
        edge_cases,
        complexity,
        critical_issues: json_string_vec(data, "critical_issues"),
        suggested_improvements: json_string_vec(data, "suggested_improvements"),
        refined_code,
        raw_response: raw,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn approved_verdict() {
        let data = serde_json::json!({
            "verdict": "APPROVED",
            "score": 95,
            "summary": "All tests passed",
            "edge_cases_identified": [{
                "case": "Empty array",
                "risk_level": "LOW",
                "handled_properly": true
            }]
        });
        let res = audit_from_value(&data, String::new());
        assert!(res.is_approved());
        assert_eq!(res.score, 95);
        assert_eq!(res.edge_cases.len(), 1);
    }
}
